Basic implementation of interval trees based on AVL tree.

rednum opened the pull request:
Hi,

this is a pull request for basic implementation of interval trees. Currently it supports two operations:
- insert
- find intersection
both of those should run in O(log N). I hope the API (generic with respect to interval bounds) is reasonable, if you have some ideas for improvement I'd be happy to hear about that.

johanneskoester commented on LICENSE.md:
```diff
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2016 Johannes Köster, the Rust-Bio team.
+Copyright (c) 2016 Johannes Köster, the Rust-Bio team, Google Inc.
```
So, is this official google work? Are you part of google? If so, I'm excited to hear that google is interested in this project!!

rednum replied:
Hi, so how this works:
- yes, I'm an engineer at Google
- no, this is not a part of my daily work at Google, rather a fun side project for me. However, Google gets copyright to open source patches I submit when employed there.

Hope this works for you?

johanneskoester replied:
Since this is MIT licensed it shouln't be a problem. Thanks for clarifying!

johanneskoester commented on src/data_structures/interval_tree.rs:
```diff
@@ -0,0 +1,315 @@
+        }
+    }
+
+    pub fn find_intersection(&self, start: N, end: N) -> Option<&D> {
```
What do you think about returning a reference to the node here? This would allow to calculate the overlap (with whatever metric comes to the users mind). Also, what happens if there are multiple intersecting intervals? Shouln't this return an Iterator so people can choose whether they want exhaustive search or the first?

rednum replied:
Yes, reference sounds reasonable. When there are multiple intervals, an arbitrary one would be returned. I think returning all intersections is a good idea, I was also thinking about this but didn't code that yet. As of returning Node, there's much tree metadata there user shouldn't really care about, so maybe something like a (start: N, end: N, data: D) tuple?

johanneskoester replied:
Makes sense. Lets iterate over (Range<N>, D) though.

johanneskoester commented on src/data_structures/interval_tree.rs:
```diff
@@ -0,0 +1,315 @@
+pub struct Node<N, D> {
+    // actual interval data
+    start: N,
+    end: N,
```
I also though about this regarding other parts of Rust-Bio recently: would you consider it a good idea to represent intervals as `Range<N>`? (a) this makes the semantic of `end` clear (inclusive or exclusive), (b) this allows intervals from here to be immediately used in slicing operations.

rednum replied:
Sounds like a good idea. However, what should be done with range objects [start..end) where start >= end? Rust doesn't prohibit from constructing such, so I suppose the following should panic:
t.insert((2..2));
t.insert((3..0));

johanneskoester replied:
Yes, I agree.

johanneskoester commented:
Great work! I have added some comments!

rednum commented:
I've added most of the discussed changes. However, I'm not sure about returning iterator when intersecting. It seems that ideally we'd want something like a lazy, streaming iterator... and it seems that streaming iterators are problematic for some reason. So for now I've added find_all, which returns a collection; I hope to read up on this in next few days.

johanneskoester commented on src/data_structures/interval_tree.rs:
```diff
@@ -0,0 +1,359 @@
+        }
+    }
+
+    pub fn find_all(&self, interval: &Range<N>) -> Vec<(&Range<N>, &D)> {
```
Maybe yield  structs `Entry` here. This is easier to use, e.g. via `entry.interval` and `entry.data`.

johanneskoester commented:
Yes, if there already was a yield statement, turning this into an iterator would be easy. Without, you would have to implement the iterator interface yourself. In this case, you would need to use a stack (e.g. with a [VecDeque](https://www.google.nl/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&cad=rja&uact=8&ved=0ahUKEwji07ikp9zOAhXEDcAKHSkdBV0QFggeMAA&url=https%3A%2F%2Fdoc.rust-lang.org%2Fstd%2Fcollections%2Fstruct.VecDeque.html&usg=AFQjCNF93rY_F1z6i0NCIAniALc7w2sK5g)) to simulate the recursion.

Let me know if you need further help.

rednum commented:
Added the iterator inteface. Unfortunately, it seems to me that the current implementation runs in pessimistic complexity O(tree size) instead of O(number of intersections). I suppose this could be improved though.

johanneskoester commented:
Without having looked at the code yet: is that a result of using the iterator or was it the same before?

rednum commented:
It was the same before adding iterator, I just didn't realize this. After the commit I've just pushed it should be O(k * log(n)^2) (k - number of intersections, n - tree size). Hopefully I'll add some benchmarks/better tests next week.

johanneskoester commented:
Sounds good, thanks!

rednum commented:
I'd like to add some benchmarks/tests, but sadly it doesn't seem like I'll have to time for this soon. If you think it's already good to merge, please do so!

johanneskoester commented:
Ok, thanks!
